x/bsonx/bsoncore/value_test.go

```diff
@@ -677,163 +677,262 @@ func TestValue(t *testing.T) {
+			val: Value{
+				Type: TypeCodeWithScope,
+				Data: AppendCodeWithScope(nil, "code",
+					BuildDocument(nil, AppendStringElement(nil, "key", "value")),
+				),
+			},
+			want: `{"$code":code,"$scope":{"key": "value"}}`,
+		},
 
-		{"n>0, TypeCodeWithScope", 10, Value{
-			Type: TypeCodeWithScope, Data: AppendCodeWithScope(nil, "code", BuildDocument(nil,
-				AppendStringElement(nil, "key", "value")))}, `{"$code":c`},
+		{
+			description: "TypeInt32",
+			val: Value{
+				Type: TypeInt32,
+				Data: AppendInt32(nil, 123),
+			},
+			want: `{"$numberInt":"123"}`,
+		},
 
-		{"n>0, TypeInt32", 10, Value{
-			Type: TypeInt32, Data: AppendInt32(nil, 123)}, `{"$numberI`},
+		{
+			description: "TypeTimestamp",
+			val: Value{
+				Type: TypeTimestamp,
+				Data: AppendTimestamp(nil, 123, 456),
+			},
+			want: `{"$timestamp":{"t":123,"i":456}}`,
+		},
 
-		{"n>0, TypeTimestamp", 10, Value{
-			Type: TypeTimestamp, Data: AppendTimestamp(nil, 123, 456)}, `{"$timesta`},
+		{
+			description: "TypeInt64",
+			val: Value{
+				Type: TypeInt64,
+				Data: AppendInt64(nil, 1234567890),
+			},
+			want: `{"$numberLong":"1234567890"}`,
+		},
 
-		{"n>0, TypeInt64", 10, Value{
-			Type: TypeInt64, Data: AppendInt64(nil, 1234567890)}, `{"$numberL`},
+		{
+			description: "TypeDecimal128",
+			val: Value{
+				Type: TypeDecimal128,
+				Data: AppendDecimal128(nil, 0x3040000000000000, 0x0000000000000000),
+			},
+			want: `{"$numberDecimal":"0"}`,
+		},
 
-		{"n>0, TypeDecimal128", 10, Value{
-			Type: TypeDecimal128, Data: AppendDecimal128(nil, 0x3040000000000000, 0x0000000000000000)}, `{"$numberD`},
+		{
+			description: "TypeMinKey",
+			val: Value{
+				Type: TypeMinKey,
+			},
+			want: `{"$minKey":1}`,
+		},
 
-		{"n>0, TypeMinKey", 10, Value{
-			Type: TypeMinKey}, `{"$minKey"`},
+		{
+			description: "TypeMaxKey",
+			val: Value{
+				Type: TypeMaxKey,
+			},
+			want: `{"$maxKey":1}`,
+		},
+	}
 
-		{"n>0, TypeMaxKey", 10, Value{
-			Type: TypeMaxKey}, `{"$maxKey"`},
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("String %s", tc.description), func(t *testing.T) {
+			got := tc.val.String()
+			assert.Equal(t, tc.want, got)
+		})
 	}
 
 	for _, tc := range testCases {
-		t.Run(tc.description, func(t *testing.T) {
-			got := tc.val.StringN(tc.n)
+		for n := -1; n <= len(tc.want)+1; n++ {
+			t.Run(fmt.Sprintf("StringN %s n==%d", tc.description, n), func(t *testing.T) {
+				got, _ := tc.val.StringN(n)
+				l := n
+				if l < 0 {
+					l = 0
+				}
+				if l > len(tc.want) {
+					l = len(tc.want)
+				}
+				want := tc.want[:l]
+				assert.Equal(t, want, got, "got %v, want %v", got, want)
+			})
+		}
+	}
+
+	multiByteString := Value{
+		Type: TypeString,
+		Data: AppendString(nil, "𨉟呐㗂越"),
+	}
+	for _, tc := range []struct {
+		n    int
+		want string
+	}{
+		{6, `"𨉟`},
+		{8, `"𨉟`},
+		{10, `"𨉟呐`},
+		{15, `"𨉟呐㗂越"`},
+		{21, `"𨉟呐㗂越"`},
+	} {
+		t.Run(fmt.Sprintf("StringN multi-byte string n==%d", tc.n), func(t *testing.T) {
```

Comment:
Optional: Consider factoring this into a separate test function `TestArray_StringN_Multibyte`.